### PR TITLE
pkg/trace/obfuscate: optimize the SQL obfuscator

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

Previously, the SQL obfuscator would build a new byte array with
bytes.Buffer for every token. This resulted in a lot of allocations. The
new obfuscator returns slices into the original query rather than new
allocations, saving a lot of work. Comparing to the previous obfuscator,
the new one is significantly better in both space and time.
```
% benchcmp master.bench new.bench
benchmark                                      old ns/op     new ns/op     delta
BenchmarkObfuscateSQLString/query1/3694-8     119274        76567         -35.81%
BenchmarkObfuscateSQLString/query3/980-8      34102         20726         -39.22%
BenchmarkObfuscateSQLString/Escaping/512-8     12644         8888          -29.71%
BenchmarkObfuscateSQLString/Grouping/199-8     6353          4858          -23.53%
BenchmarkObfuscateSQLString/random-8           1929          1378          -28.56%

benchmark                                      old allocs     new allocs     delta
BenchmarkObfuscateSQLString/query1/3694-8     912            83             -90.90%
BenchmarkObfuscateSQLString/query3/980-8      242            6              -97.52%
BenchmarkObfuscateSQLString/Escaping/512-8     93             12             -87.10%
BenchmarkObfuscateSQLString/Grouping/199-8     74             10             -86.49%
BenchmarkObfuscateSQLString/random-8           24             8              -66.67%

benchmark                                      old bytes     new bytes     delta
BenchmarkObfuscateSQLString/query1/3694-8     43489         16256         -62.62%
BenchmarkObfuscateSQLString/query3/980-8      11264         3008          -73.30%
BenchmarkObfuscateSQLString/Escaping/512-8     3264          1936          -40.69%
BenchmarkObfuscateSQLString/Grouping/199-8     1920          816           -57.50%
BenchmarkObfuscateSQLString/random-8           752           248           -67.02%
```

### Motivation

The SQL obfuscator has been the source of frequent performance issues. 

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

The test suite passes, but since this is a large change we probably want to do some testing against real applications.

### CODE REVIEW

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 1]
```diff
@@ -246,198 +247,210 @@ func (tkn *SQLTokenizer) skipBlank() {
 	for unicode.IsSpace(tkn.lastChar) {
 		tkn.next()
 	}
+	tkn.grab()
+}
+
+// myFastToUpper returns a copy of the byte slice s with all Unicode letters mapped to
+// their upper case. It uses the array pointed to by dst as the destination if possible.
+func myFastToUpper(src, dst []byte) []byte {
```
Comment (clutchski) [inline pkg/trace/obfuscate/sql_tokenizer.go:255]: I think this method name is a bit rough. Just call it `uppercase` and explain why its fast in a comment.

**Comment by clutchski** [comment]
Obviously testing is important as you mentioned. Great work. I like the attention to low level details.

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 2]
```diff
@@ -246,198 +247,210 @@ func (tkn *SQLTokenizer) skipBlank() {
...
-
 	for isLetter(tkn.lastChar) || isDigit(tkn.lastChar) || tkn.lastChar == '.' || tkn.lastChar == '*' {
-		buffer.WriteRune(tkn.lastChar)
 		tkn.next()
 	}
-	upper := bytes.ToUpper(buffer.Bytes())
+
+	t := tkn.grab()
+	// Space allows us to upper-case identifiers 200 bytes long or less without allocating.
+	var space [200]byte
```
Comment (clutchski) [inline pkg/trace/obfuscate/sql_tokenizer.go:297]: why 200?

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 3]
```diff
@@ -246,198 +247,210 @@ func (tkn *SQLTokenizer) skipBlank() {
...
+		}
+		hasLower = hasLower || ('a' <= c && c <= 'z')
+	}
+	if cap(dst) < len(src) {
+		dst = make([]byte, 0, len(src))
+	}
+	if isASCII { // optimize for ASCII-only byte slices.
+		if !hasLower {
+			// Just return a copy.
+			return append(dst, src...)
```
Comment (clutchski) [inline pkg/trace/obfuscate/sql_tokenizer.go:272]: i think in this case can't we just use the original string?

File: pkg/trace/obfuscate/sql.go [hunk 4]
```diff
@@ -265,20 +265,17 @@ func (oq *ObfuscatedQuery) Cost() int64 {
...
-		&groupingFilter{},
-	}
 	tableFinder := &tableFinderFilter{}
-	if config.HasFeature("table_names") {
-		filters = append(filters, tableFinder)
-	}
+	filterTableFinder := config.HasFeature("table_names")
 	var (
-		out       bytes.Buffer
+		out       bytes.Buffer = *bytes.NewBuffer(make([]byte, 0, len(tokenizer.buf)))
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql.go:271]:
```suggestion
		out       = bytes.NewBuffer(make([]byte, 0, len(tokenizer.buf)))
```
Would this work?

File: pkg/trace/obfuscate/sql.go [hunk 5]
```diff
@@ -265,20 +265,17 @@ func (oq *ObfuscatedQuery) Cost() int64 {
 // attemptObfuscation attempts to obfuscate the SQL query loaded into the tokenizer, using the
 // given set of filters.
 func attemptObfuscation(tokenizer *SQLTokenizer) (*ObfuscatedQuery, error) {
-	filters := []tokenFilter{
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql.go:268]: Why did you change this logic? Did it bring any benefit?

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 6]
```diff
@@ -80,10 +79,11 @@ const escapeCharacter = '\\'
...
 // tokens for the parser.
 type SQLTokenizer struct {
-	rd       *strings.Reader // the "rune" reader
-	pos      int             // byte offset of lastChar
-	lastChar rune            // last read rune
-	err      error           // any error occurred while reading
+	pos      int    // byte offset of lastChar
+	lastChar rune   // last read rune
+	buf      []byte // buf holds the query that we are parsing
+	index    int    // index is the index into buf where the unread portion of the query begins.
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_tokenizer.go:85]: This is generally called `off` (reading offset) https://golang.org/src/bytes/buffer.go#L22. Maybe it'd be nice to use the same since we're doing something similar?

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 1 above)
Comment (gbbr) [reply pkg/trace/obfuscate/sql_tokenizer.go:255]: When copying code from somewhere, let's make sure we always mention where it is coming from. Along with that an explanation of why the standard library `bytes.ToUpper` is mandatory IMHO. It will help with both understanding and maintaining it.

Lastly, how much more benefit did this bring?

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 7]
```diff
@@ -480,60 +493,68 @@ func (tkn *SQLTokenizer) scanString(delim rune, kind TokenKind) (TokenKind, []by
...
-	return Comment, buffer.Bytes()
+	return Comment, tkn.grab()
 }
 
-func (tkn *SQLTokenizer) consumeNext(buffer *bytes.Buffer) {
-	if tkn.lastChar == EOFChar {
-		// This should never happen.
-		panic("unexpected EOF")
+// Next sets tkn.lastChar to the next rune at tkn.buf[tkn.index], and advances the tkn.pos and
+// tkn.index by its size. If next() is unable to scan a character, tkn.lastChar is set to EOFChar.
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_tokenizer.go:526]: The documentation is a bit too "code like". Maybe it should just say
```go
// next advances the tokenizer to the next rune. If this is not possible, tkn.lastChar will
// be set to EOFChar.
``` 
The technical details are in the code and the scope of the fields are in the structure definition comments.

Since this is called a "tokenizer", it's easy to think that `next` takes us to the next token, so maybe `nextRune` might be better, or `advance`

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 8]
```diff
@@ -480,60 +493,68 @@ func (tkn *SQLTokenizer) scanString(delim rune, kind TokenKind) (TokenKind, []by
...
-		tkn.lastChar = ch
+		tkn.pos += len
 	}
+	tkn.index += len
+	tkn.lastChar = ch
+}
+
+// grab returns a byte slice of all the characters advanced over with next() since the last call to
+// grab.
+func (tkn *SQLTokenizer) grab() []byte {
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_tokenizer.go:547]: Could `bytes` be a good name for this method?
```go
// bytes returns all the bytes that were advanced over since its last call.
```

**Review by gbbr — COMMENTED** [review]
This is really good Kyle! Let's make sure it's tested properly before merging, and feel free to ask for another review after you touch everything up.

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 1 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:255]: Sorry, I meant to rename this before I pushed it, but missed it. 
I will give it a more appropriate name and document why we want to use this instead of the standard library one.

I can put up a benchmark with/without it. It saves an allocation on every identifier.

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 6 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:85]: 👍

File: pkg/trace/obfuscate/sql.go (hunk 5 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql.go:268]: Yes, when profiling, this showed up as a significant source of allocations - allocating an array, several objects, and then interfaces for them for each query.

I didn't record exactly how much this specific change improved performance, but I thought it was worthwhile.

File: pkg/trace/obfuscate/sql.go (hunk 4 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql.go:271]: 👍

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 3 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:272]: 👍  Yes, we can.

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 2 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:297]: 200 was chosen because it seemed reasonable. We may be able to find a more appropriate number. This is stack allocated as you mentioned, so we can get a larger array for the same (negligible) cost if we want. I'm not sure how common identifiers larger than 200 bytes are.

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 8 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:547]: `bytes` will be better I think.

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 7 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:526]: 👍  I will simplify the comment.
I like `advance` as a name.

File: pkg/trace/obfuscate/sql.go (hunk 5 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql.go:268]: The overhead ends up being pretty significant.

I ran this benchmark against attemptObfuscation using the filter slice method (`oldattempt.txt`) and the one in this PR (`newattempt.txt`). using `go test -v -run NONE -bench BenchmarkObfuscateSQLString -benchmem -benchtime 30s`
Everything else was the same in both (including the `out = ` change)
```
% benchcmp oldattempt.txt newattempt.txt
benchmark                                      old ns/op     new ns/op     delta
BenchmarkObfuscateSQLString/Escaping/512-8     9594          7894          -17.72%
BenchmarkObfuscateSQLString/Grouping/199-8     4945          3791          -23.34%
BenchmarkObfuscateSQLString/query1/3694-8      83517         69018         -17.36%
BenchmarkObfuscateSQLString/query3/980-8       23150         17440         -24.67%
BenchmarkObfuscateSQLString/random-8           1596          1157          -27.51%

benchmark                                      old allocs     new allocs     delta
BenchmarkObfuscateSQLString/Escaping/512-8     25             12             -52.00%
BenchmarkObfuscateSQLString/Grouping/199-8     24             10             -58.33%
BenchmarkObfuscateSQLString/query1/3694-8      152            83             -45.39%
BenchmarkObfuscateSQLString/query3/980-8       13             6              -53.85%
BenchmarkObfuscateSQLString/random-8           11             8              -27.27%

benchmark                                      old bytes     new bytes     delta
BenchmarkObfuscateSQLString/Escaping/512-8     2011          1936          -3.73%
BenchmarkObfuscateSQLString/Grouping/199-8     892           816           -8.52%
BenchmarkObfuscateSQLString/query1/3694-8      16387         16256         -0.80%
BenchmarkObfuscateSQLString/query3/980-8       3077          3008          -2.24%
BenchmarkObfuscateSQLString/random-8           320           248           -22.50%
```

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 1 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:255]: Here is the benchmark for `bytes.ToUpper` vs `myFastToUpper`. `newToUpper` is this PR. `oldToUpper` is this PR plus the following diff:
```diff
--- a/pkg/trace/obfuscate/sql_tokenizer.go
+++ b/pkg/trace/obfuscate/sql_tokenizer.go
@@ -294,8 +294,9 @@ func (tkn *SQLTokenizer) scanIdentifier() (TokenKind, []byte) {
 
 	t := tkn.grab()
 	// Space allows us to upper-case identifiers 200 bytes long or less without allocating.
-	var space [200]byte
-	upper := myFastToUpper(t, space[:0])
+	//var space [200]byte
+	//upper := myFastToUpper(t, space[:0])
+	upper := bytes.ToUpper(t)
 	if keywordID, found := keywords[string(upper)]; found {
 		return keywordID, t
 	}
```

```
% benchcmp oldToUpper.txt newToUpper.txt 
benchmark                                      old ns/op     new ns/op     delta
BenchmarkObfuscateSQLString/Escaping/512-8     8809          7652          -13.13%
BenchmarkObfuscateSQLString/Grouping/199-8     4208          3780          -10.17%
BenchmarkObfuscateSQLString/query1/3694-8      78783         67797         -13.94%
BenchmarkObfuscateSQLString/query3/980-8       20154         18383         -8.79%
BenchmarkObfuscateSQLString/random-8           1220          1305          +6.97%

benchmark                                      old allocs     new allocs     delta
BenchmarkObfuscateSQLString/Escaping/512-8     32             12             -62.50%
BenchmarkObfuscateSQLString/Grouping/199-8     21             10             -52.38%
BenchmarkObfuscateSQLString/query1/3694-8      405            83             -79.51%
BenchmarkObfuscateSQLString/query3/980-8       96             6              -93.75%
BenchmarkObfuscateSQLString/random-8           13             8              -38.46%

benchmark                                      old bytes     new bytes     delta
BenchmarkObfuscateSQLString/Escaping/512-8     2144          1936          -9.70%
BenchmarkObfuscateSQLString/Grouping/199-8     928           816           -12.07%
BenchmarkObfuscateSQLString/query1/3694-8      20336         16256         -20.06%
BenchmarkObfuscateSQLString/query3/980-8       4224          3008          -28.79%
BenchmarkObfuscateSQLString/random-8           288           248           -13.89%
```

For renaming, I chose `toUpper` to be consistent with its source, `bytes.ToUpper`.

File: pkg/trace/obfuscate/sql_tokenizer.go (hunk 2 above)
Comment (knusbaum) [reply pkg/trace/obfuscate/sql_tokenizer.go:297]: I did a bit of research, and I think 256 bytes is a good number.
* Identifiers in mysql can be up to [256 characters](https://dev.mysql.com/doc/refman/8.0/en/identifier-length.html) (but only aliases on table names. Most are limited to 64 characters)
* For postgres, it's [63 characters](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS)
* For Oracle it's [128 characters](https://oracle-base.com/articles/12c/long-identifiers-12cr2#:~:text=In%20Oracle%20Database%2012cR2%2C%20the,identifiers%20is%20now%20128%20characters.)
* For MS SQL Server, the limit (as best I can tell) is [128 characters](https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2008-r2/ms191240(v=sql.105)?redirectedfrom=MSDN#sysname)

As far as I can tell, ANSI/ISO SQL:2011 only specifies at least 30 characters.

256 bytes will cover almost all cases.
Characters and bytes aren't equivalent, but I think it's OK if queries with long, utf-filled aliases in mysql hit an extra alloc.

File: pkg/trace/obfuscate/sql.go [hunk 9]
```diff
@@ -290,8 +287,18 @@ func attemptObfuscation(tokenizer *SQLTokenizer) (*ObfuscatedQuery, error) {
...
+		if token, buff, err = discard.Filter(token, lastToken, buff); err != nil {
+			return nil, err
+		}
+		if token, buff, err = replace.Filter(token, lastToken, buff); err != nil {
+			return nil, err
+		}
+		if token, buff, err = grouping.Filter(token, lastToken, buff); err != nil {
+			return nil, err
+		}
+		if filterTableFinder {
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql.go:300]:
```suggestion
		if config.HasFeature("table_names") {
```
I guess we can avoid that variable.

File: pkg/trace/obfuscate/sql_test.go [hunk 10]
```diff
@@ -1015,6 +1086,8 @@ func BenchmarkObfuscateSQLString(b *testing.B) {
 	}{
 		{"Escaping", `INSERT INTO delayed_jobs (attempts, created_at, failed_at, handler, last_error, locked_at, locked_by, priority, queue, run_at, updated_at) VALUES (0, '2016-12-04 17:09:59', NULL, '--- !ruby/object:Delayed::PerformableMethod\nobject: !ruby/object:Item\n  store:\n  - a simple string\n  - an \'escaped \' string\n  - another \'escaped\' string\n  - 42\n  string: a string with many \\\\\'escapes\\\\\'\nmethod_name: :show_store\nargs: []\n', NULL, NULL, NULL, 0, NULL, '2016-12-04 17:09:59', '2016-12-04 17:09:59')`},
 		{"Grouping", `INSERT INTO delayed_jobs (created_at, failed_at, handler) VALUES (0, '2016-12-04 17:09:59', NULL), (0, '2016-12-04 17:09:59', NULL), (0, '2016-12-04 17:09:59', NULL), (0, '2016-12-04 17:09:59', NULL)`},
+		{"query1", query1},
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_test.go:1089]:
```suggestion
		{"Large", query1},
```

File: pkg/trace/obfuscate/sql_test.go [hunk 11]
```diff
@@ -1015,6 +1086,8 @@ func BenchmarkObfuscateSQLString(b *testing.B) {
...
+		{"query3", query3},
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_test.go:1090]:
```suggestion
		{"Complex", query3},
```
To keep a bit of uniformity.

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 12]
```diff
@@ -230,225 +231,244 @@ func (tkn *SQLTokenizer) Scan() (TokenKind, []byte) {
...
+		}
+		hasLower = hasLower || ('a' <= c && c <= 'z')
+	}
+	if cap(dst) < len(src) {
+		dst = make([]byte, 0, len(src))
+	}
+	if isASCII { // optimize for ASCII-only byte slices.
+		if !hasLower {
+			// Just return a copy.
+			//return append(dst, src...)
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_tokenizer.go:277]: What's this?

File: pkg/trace/obfuscate/sql_tokenizer.go [hunk 13]
```diff
@@ -230,225 +231,244 @@ func (tkn *SQLTokenizer) Scan() (TokenKind, []byte) {
...
+		}
+		dst = dst[:len(src)]
+		//b := make([]byte, len(src))
```
Comment (gbbr) [inline pkg/trace/obfuscate/sql_tokenizer.go:280]: Does this help?

**Review by gbbr — COMMENTED** [review]
Mostly looks good now but you've left some commented out code.

**Comment by knusbaum** [comment]
Cleaned up and added another test for our new `toUpper`

File: releasenotes/notes/refactor-sql-obfuscator-3c8a0d1ef428c410.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    APM: refactor the SQL obfuscator to be significantly more efficient.
```
Comment (gbbr) [inline releasenotes/notes/refactor-sql-obfuscator-3c8a0d1ef428c410.yaml:11]:
```suggestion
    APM: refactored the SQL obfuscator to be significantly more efficient.
```

**Review by gbbr — APPROVED** [review]
This is excellent! Thank you Kyle!

File: pkg/trace/obfuscate/sql.go (hunk 9 above)
Comment (gbbr) [reply pkg/trace/obfuscate/sql.go]: Good point. I think that's a sign that we should make `config.HasFeature` more efficient. Perhaps by loading the features only once at load time.
